### PR TITLE
Rewrote package to use `composer-runtime-api:^2.0` and no longer keep an internal generated map

### LINKED ISSUE #90: Test psalm annotations against a happy path/unhappy path static analysis test suite
TODOs:

 * [ ] https://github.com/vimeo/psalm/issues/1881

### LINKED ISSUE #116: FallbackVersions uses this package lock which breaks version checks
In some cases `PackageVersions\Versions`  does not have version for package being checked:

* In case of package `replaces` and `provides` (i.e symfony/symfony replaces symfony/process)
* In case of `--no-scripts` was used while doing `composer install` (for security reasons)

These are different cases which exposes differently, but the root cause is `FallbackVersions` using this package `composer.lock` file. It does not filter out `dev` packages.

https://github.com/infection/infection/issues/1007

https://github.com/scaytrase/package-version-fallback-issue
https://travis-ci.org/scaytrase/package-version-fallback-issue/builds/639467674

In linked travis build you can examine that `infection/infection` thinks it runs `0.13.4` (while required, locked and installed as `0.15.0`, see logs). But this is minor issue.
The major one is that infection thinks that it uses symfony/process 4.3.2 from this package lock file, but this is not true, since `symfony/symfony` is required and installed as ^5.0 and doesn't have called method anymore (see travis logs)

I didn't know what are the reasons of including this package lock into the FallbackVersions, but my suggested fix is to drop them out of there.

IMO I'd better get `no package installed` error and check possible fix ways instead of silently getting wrong version from fallback

### LINKED ISSUE #135: Drop support for Composer 1.x once stable
See #133

The added code should be dropped once stable `composer/composer:2.0.0` is out.

### PR DESCRIPTION
**Pull request by Ocramius**
This massively simplifies the whole design, as it allows us to get rid of most of the complexity
of the package. Sadly, due to the API of `Versions::ROOT_PACKAGE_NAME`, we still need to generate
code during installation, and the composer plugin needs to be maintained long-term, but overall
we can deprecate its usage, and endorse moving to `composer/composer:^2.0.0`.

### CODE REVIEW

File: src/PackageVersions/Installer.php [hunk 1]
```diff
@@ -48,15 +49,13 @@ final class Installer implements ComposerV2Plugin, EventSubscriberInterface
...
-     * Dont read this array from your calling code, but use the \PackageVersions\Versions::getVersion() method instead.
-     *
-     * @var array<string, string>
-     * @internal
+     * @deprecated please use {@see \Composer\InstalledVersions::getRootPackage()} instead. The
+     *             equivalent expression for this constant's contents is
+     *             `\Composer\InstalledVersions::getRootPackage()['name']`.
+     *             This constant will be removed in version 2.0.0.
      */
-    public const VERSIONS          = %s;
```
Comment (alexander-schranz) [inline src/PackageVersions/Installer.php:59]: @ocramius It seems like removing this seems to break compatibility to [ocramius/proxy-manager:2.8](https://github.com/Ocramius/ProxyManager/blob/2.8.0/src/ProxyManager/Generator/Util/IdentifierSuffixer.php#L50)

I get the following error:

> !!  Fatal error: Uncaught Error: Undefined class constant 'VERSIONS' in /tmp/sulu-develop.localhost/vendor/ocramius/proxy-manager/src/ProxyManager/Generator/Util/IdentifierSuffixer.php:50

Comment (Ocramius) [reply src/PackageVersions/Installer.php:59]: Fairly sure that's my own fault inside `ocramius/proxy-manager` then: can you report an issue there, please?

Comment (Ocramius) [reply src/PackageVersions/Installer.php:59]: Note: this was `@internal` before, so I shouldn't have relied on it in first place.

Comment (alexander-schranz) [reply src/PackageVersions/Installer.php:59]: Oh I see make sense then that it need to be changed at the proxy-manager side.

**Pull request merged**